// s3-deploy-plugin/index.js
'use strict';

const chalk = require('chalk');
const s3 = require('s3');

class S3DeployPlugin {
  
  constructor() {
      
    this.commands = {
      deploy: {
        commands: {
          frontend: {
            usage: 'Deploy frontend to S3',
            lifecycleEvents: [
              'deploy-s3'
            ],
          },
        },
      },
    };  
      
    this.hooks = {
      'deploy:frontend:deploy-s3': this.afterDeployDeployToS3,
      'after:deploy:deploy': this.afterDeployDeployToS3
    };
  }

  afterDeployDeployToS3() {
	var s3client = s3.createClient();
	var s3params = {
	  localDir: "frontend",
	  deleteRemoved: true, 
	  s3Params: {
	    Bucket: "dev-demostore-frontend",
	    Prefix: "",
	    ACL: 'public-read'
	  },
	};
	var uploader = s3client.uploadDir(s3params);
	
    console.log(`Endless: ${chalk.yellow('Uploading frontend to S3...')}`);
	uploader.on('progress', function() {
	  process.stdout.write(`${chalk.yellow('.')}`);
	});
	
	uploader.on('end', function() {
	  let url = s3.getPublicUrlHttp("dev-demostore-frontend",'index.html');		
	  let message = `
	  
${chalk.yellow.underline('Endless Information')}
${chalk.yellow('frontend:')} ${url}`;
	  console.log(message);
	});

  }
}

module.exports = S3DeployPlugin;
